File: Java/BOJ/2628 종이자르기.java
import java.io.*;
import java.util.*;
import java.math.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int T = Integer.parseInt(br.readLine());
        
        ArrayList<Integer> w = new ArrayList<>();
        ArrayList<Integer> h = new ArrayList<>();

        w.add(0);
        w.add(n);
        
        h.add(0);
        h.add(m);

        int result= 0;
        for (int i=0; i<T; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int line = Integer.parseInt(st.nextToken());

            if (x == 0) {
                h.add(line);
            } else {
                w.add(line);
            }
            
            Collections.sort(w);
            Collections.sort(h);

       }
        for (int j=0; j<w.size()-1; j++) {
        	for (int k=0; k<h.size()-1; k++) {
        		
        		int x_diff=w.get(j+1)-w.get(j);
        		int y_diff=h.get(k+1)-h.get(k);
        		
        		result=Math.max(result, x_diff*y_diff);  
        	}
        }
        System.out.println(result); 
    }
}

File: Python/BOJ/2628 종이자르기.py
import sys
n, m = map(int, sys.stdin.readline().split())
T = int(sys.stdin.readline())
w, h = [0, n], [0, m]
for _ in range(T):
    x, line = map(int,sys.stdin.readline().split())
    if x == 0:
        h.append(line)
    else:
        w.append(line)

w.sort()
h.sort()

result = 0
for i in range(len(w) -1):
    for j in range(len(h) - 1):
        x = w[i+1] - w[i]
        y = h[j+1] - h[j]
        result = max(result, x*y)
print(result)
